### ISSUE TITLE
Dynamically connecting Python and C# clients to TraCI

### ISSUE THREAD

**Issue by MohamadReza-Haghani** [eclipse/sumo#12730]
Hi,

I am working on connecting Unity to SUMO using the C# port of TraCI(@CodingConnected). I also intend to add a Python TraCI connection to SUMO to control the simulation using both Python and C#. I tried starting SUMO with `--num-clients <INT>` to make multiple connections.
Right now, C# connects to TraCI as the primary connection. Then, the Python script is called from the C# script and connects to TraCI as the second connection. The problem is that Unity does not allow the Python script to continue along with the C# script. Unity waits until the Python script is finished and, then, moves forward to the next simulation step. Therefore, I need to write a Python script that only performs one simulation step. Therefore, C# calls this Python script in every simulation step.
 
My first question is whether the Python connection to TraCI can be saved as a variable somewhere. If so, Python does not close its connection, saves the connection as a variable, and connects to the same open connection in the next call from C#. 
My second question is whether Python can close its connection in every simulation step of Unity and connect again in the next simulation step while C# does not close its connection until the end of the simulation. 
I read that for the dynamic connection of multiple clients, I need to write my own proxy and control the connections with it. If this is the solution, I would appreciate it if you guide me on how to do it.

Thanks,
Reza.

**Comment by namdre** [comment]
currently, multi-client support assumes that the same clients all connect at the start and stay connected until simulation end. We have plans to change this (#6672)  but it's just not possible at the moment.

Also, the repo at https://github.com/CodingConnected/CodingConnected.Traci hasn't seen any maintenance since 2021 and has known problems: #10748,  #10875, 

The recommended alternative is to use the libsumo C# bindings: #11712.
Even though the latter are somewhat experimental, there will be support from the main team.

**Comment by namdre** [comment]
My suggestion would be to find a way for unity to launch your python script in the background rather than waiting for it to finish.

**Comment by MohamadReza-Haghani** [comment]
Thank you for your answer. I found a way to call Python code in the back ground and keep it running. When I call the Python code using command prompt, it runs in the background.

Right now, when I try to make a single connection to TraCI from either C# using [https://github.com/CodingConnected/CodingConnected.Traci](url) or from Python, it works correctly. But when I try to connect TraCI to multiple clients according the example shown in [https://sumo.dlr.de/docs/TraCI/Interfacing_TraCI_from_Python.html](url), it gives me the following error in command prompt:
`traci.init(port=4001, numRetries=60, host="127.0.0.1", label="default", proc=None, doSwitch=True)
  File "C:\Program Files (x86)\Eclipse\Sumo\tools\traci\main.py", line 140, in init
    return _connections[label].getVersion()
  File "C:\Program Files (x86)\Eclipse\Sumo\tools\traci\connection.py", line 369, in getVersion
    result = self._sendCmd(command, None, None)
  File "C:\Program Files (x86)\Eclipse\Sumo\tools\traci\connection.py", line 181, in _sendCmd
    return self._sendExact()
  File "C:\Program Files (x86)\Eclipse\Sumo\tools\traci\connection.py", line 91, in _sendExact
    raise FatalTraCIError("connection closed by SUMO")
traci.exceptions.FatalTraCIError: connection closed by SUMO`

I first open up SUMO in command prompt:
`sumo-gui -c map.sumocfg --start --num-clients 2 --remote-port 4001 --step-length 0.02 --quit-on-end`

Then, I make connection to TraCI as the first client in C#:
`public TraCIClient client;
client = new TraCIClient();
client.Connect("127.0.0.1", 4001); //connects to SUMO simulation
client.Control.SetOrder(1);`

At the end, I open up the Python script using command prompt and make the second connection in Python:
`traci.init(port=4001, numRetries=60, host="127.0.0.1", label="default", proc=None, doSwitch=True)
traci.setOrder(2)`

I looked into the libsumo C# bindings that you mentioned before but did not figure out how I can use libsumo in C#. Is there any example for it?

Thanks,
Reza.
